File: packages/faro-ios-sdk/Sources/FaroSDK/Transport/HttpTransport.swift
import Foundation

internal enum HttpTransportError: Error {
    case invalidCollectorUrl(String)
}

internal final class HttpTransport: Transport {
    let name = "faro-ios:transport-http"

    private let collectorUrl: URL
    private let apiKey: String?
    private let logger: InternalLogger
    private let session: URLSession
    private var disabledUntil: Date = .distantPast

    private static let tooManyRequests = 429
    private static let accepted = 202
    private static let defaultRateLimitBackoffSeconds: TimeInterval = 5.0

    init(collectorUrl: String, apiKey: String?, logger: InternalLogger) throws {
        guard let url = URL(string: collectorUrl) else {
            throw HttpTransportError.invalidCollectorUrl(collectorUrl)
        }
        self.collectorUrl = url
        self.apiKey = apiKey
        self.logger = logger

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 10
        self.session = URLSession(configuration: config)
    }

    func send(body: TransportBody) throws {
        if Date() < disabledUntil {
            logger.warn("Transport rate limited, dropping payload until \(disabledUntil)")
            return
        }

        let jsonData = try body.toJSON()

        var request = URLRequest(url: collectorUrl)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let apiKey = apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        }

        if let sessionId = body.meta.session?.id {
            request.setValue(sessionId, forHTTPHeaderField: "x-faro-session-id")
        }

        request.httpBody = jsonData

        let semaphore = DispatchSemaphore(value: 0)
        var transportError: Error?

        let task = session.dataTask(with: request) { [weak self] _, response, error in
            defer { semaphore.signal() }

            if let error = error {
                self?.logger.error("Failed to send payload", error: error)
                transportError = error
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else { return }

            switch httpResponse.statusCode {
            case Self.accepted:
                self?.logger.debug("Payload sent successfully")
            case Self.tooManyRequests:
                let backoff = self?.parseRetryAfter(httpResponse) ?? Self.defaultRateLimitBackoffSeconds
                self?.disabledUntil = Date().addingTimeInterval(backoff)
                self?.logger.warn("Rate limited, backing off for \(backoff)s")
            default:
                self?.logger.error("Unexpected response code: \(httpResponse.statusCode)")
            }
        }
        task.resume()
        semaphore.wait()

        if let error = transportError {
            throw error
        }
    }

    func shutdown() {
        session.invalidateAndCancel()
    }

    private func parseRetryAfter(_ response: HTTPURLResponse) -> TimeInterval {
        guard let retryAfter = response.value(forHTTPHeaderField: "Retry-After") else {
            return Self.defaultRateLimitBackoffSeconds
        }

        if let seconds = Double(retryAfter) {
            return seconds
        }

        let formatter = DateFormatter()
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        if let date = formatter.date(from: retryAfter) {
            return max(0, date.timeIntervalSinceNow)
        }

        return Self.defaultRateLimitBackoffSeconds
    }
}

File: packages/faro-android-sdk/src/main/kotlin/com/grafana/faro/transport/HttpTransport.kt
package com.grafana.faro.transport

import com.grafana.faro.internal.InternalLogger
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

internal class HttpTransport(
    private val collectorUrl: String,
    private val apiKey: String?,
    private val logger: InternalLogger,
    private val onRateLimited: ((retryAfterMs: Long) -> Unit)? = null
) : Transport {
    override val name = "faro-android:transport-http"

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json".toMediaType()

    companion object {
        private const val TOO_MANY_REQUESTS = 429
        private const val ACCEPTED = 202
        private const val DEFAULT_RATE_LIMIT_BACKOFF_MS = 5000L
    }

    @Volatile
    private var disabledUntilMs: Long = 0

    override fun send(body: TransportBody) {
        val now = System.currentTimeMillis()
        if (now < disabledUntilMs) {
            logger.warn("Transport rate limited, dropping payload until $disabledUntilMs")
            return
        }

        val jsonBody = TransportBody.toJson(body)

        val requestBuilder = Request.Builder()
            .url(collectorUrl)
            .post(jsonBody.toRequestBody(jsonMediaType))
            .header("Content-Type", "application/json")

        apiKey?.let {
            requestBuilder.header("x-api-key", it)
        }

        body.meta.session?.id?.let { sessionId ->
            requestBuilder.header("x-faro-session-id", sessionId)
        }

        try {
            val response = client.newCall(requestBuilder.build()).execute()
            response.use {
                when (response.code) {
                    ACCEPTED -> {
                        logger.debug("Payload sent successfully")
                    }
                    TOO_MANY_REQUESTS -> {
                        val retryAfterMs = parseRetryAfter(response.header("Retry-After"))
                        disabledUntilMs = System.currentTimeMillis() + retryAfterMs
                        logger.warn("Rate limited, backing off for ${retryAfterMs}ms")
                        onRateLimited?.invoke(retryAfterMs)
                    }
                    else -> {
                        logger.error("Unexpected response code: ${response.code}")
                    }
                }
            }
        } catch (e: Exception) {
            logger.error("Failed to send payload", e)
            throw TransportException("Failed to send payload to $collectorUrl", e)
        }
    }

    override fun shutdown() {
        client.dispatcher.executorService.shutdown()
        client.connectionPool.evictAll()
    }

    private fun parseRetryAfter(header: String?): Long {
        if (header == null) return DEFAULT_RATE_LIMIT_BACKOFF_MS

        // Try parsing as seconds
        header.toLongOrNull()?.let { return it * 1000 }

        // Try parsing as HTTP date
        try {
            val date = java.text.SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz", java.util.Locale.US)
                .parse(header)
            if (date != null) {
                return maxOf(0, date.time - System.currentTimeMillis())
            }
        } catch (_: Exception) {}

        return DEFAULT_RATE_LIMIT_BACKOFF_MS
    }
}
